learn/classes/abstract_class.md

```diff
@@ -0,0 +1,316 @@
+
+### Interfaces in Python
+
+An interface is a collection of abstract methods that defines the behavior of a class. In Python, there is no strict definition of an interface like in other programming languages like Java. Instead, we use abstract classes to define interfaces. An interface can be considered as a contract between the class and the outside world. A class that implements an interface must provide an implementation for all the methods defined in the interface.
+
+#### Example of a Python Interface
+
+```python
+from abc import ABC, abstractmethod
+ 
+class Shape(ABC):
+     
+    @abstractmethod
+    def area(self):
+        pass
+ 
+    @abstractmethod
+    def perimeter(self):
+        pass
+ 
+class Rectangle(Shape):
+     
+    def __init__(self, length, width):
+        self.length = length
+        self.width = width
+     
+    def area(self):
+        return self.length * self.width
+     
+    def perimeter(self):
+        return 2 * (self.length + self.width)
+
+r = Rectangle(10, 5)
+print("Area of Rectangle:", r.area())  # Area of Rectangle: 50
+print("Perimeter of Rectangle:", r.perimeter()) # Perimeter of Rectangle: 30
+```
+
+In the above example, Shape is an interface with two abstract methods called area() and perimeter(). Rectangle is a class that implements the Shape interface and provides its implementation for the area() and perimeter() methods.  
+  
+## Polymorphism in Python Abstract Classes  
+
+An abstract class is a class that cannot be instantiated, but it can be used as a base for other classes. It is a way to define a common interface for a group of related classes. Polymorphism is the ability of an object to take on many forms. In Python, abstract classes can be used to implement polymorphism.
+
+
+```python
+from abc import ABC, abstractmethod
+
+class Animal(ABC):
+    @abstractmethod
+    def move(self):
+        pass
+
+class Dog(Animal):
+    def move(self):
+        print("Dog is running")
+
+class Cat(Animal):
+    def move(self):
+        print("Cat is jumping")
+
+d = Dog()
+c = Cat()
+
+d.move() ### This will print "dog is running"
+c.move() ### This will print "cat is jumping"
+
+```
+  
+## Usage of Python Abstract Base Classes
+
+Abstract classes in Python are classes that cannot be instantiated, and are meant to be inherited by other classes. They are useful for defining common methods and properties that should be implemented by their concrete subclasses. 
+
+One way to implement abstract classes in Python is to use abstract base classes (ABCs) from the `abc` module. An ABC is a special type of class that cannot be instantiated directly, but can be subclassed. ABCs define one or more abstract methods, which are methods that must be implemented by any concrete subclass. 
+
+The following code example shows how to define an abstract base class `Animal`, which defines two abstract methods `speak` and `move` that must be implemented by any concrete subclass:
+
+```python
+### Example 1
```

Comment:
A lot of such mistaken headings in the code samples. Move them outside of a code sample and set as a normal heading.